# Part I - Exploring Prosper Loan Data
## by Israel Ogunmola

## Introduction
---

If you want to consolidate debt, finance a large purchase, or cover an emergency expense, a personal loan can be immensely useful. However, many personal loan lenders require borrowers to have good or excellent credit, making it difficult to qualify for a loan. Borrowers with credit ratings in the fair range or below may have better chances of obtaining loans at a better rate by working with a peer-to-peer lender.

[Prosper is a personal loan pioneer](https://www.prosper.com/) — the US company became the first firm to enter the peer-to-peer lending arena when it launched in 2005. Since then, the platfom has originated more than 20 billion USD in personal loans by matching over 1,170,000 borrowers to potential investors through its online platform. Prosper offers unsecured personal loans to customers who have a minimum credit score of 640. It also provides home equity lines of credit (HELOCs).

Our goal is to explore a sample of Prosper loan data to uncover borrower motivations when applying for loans, and identify several factors that may influence loan favorability.

## Importing Libraries
---
 
A great way to start is by importing the libraries and packages we need. We will import the **Numpy** and **Pandas** libraries to help us load and perform quick, vectorized operations on our data, then the **Matplotlib** and **Seaborn** libraries to help us build informing visuals:

In [1]:
# Data analysis and visualization packages
import numpy as np
import pandas as pd
from IPython.display import HTML, display
import matplotlib.pyplot as plt
import seaborn as sb

# Configure visualization behaviours
%matplotlib inline
# %config InlineBackend.figure_format = "retina"

## Preliminary Wrangling
---
We will start by importing our dataset, `prosperLoanData.csv`, then reading it into a pandas dataframe:

In [2]:
df = pd.read_csv('./prosperLoanData.csv')

In [3]:
# Display quick summary information about the dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 113937 entries, 0 to 113936
Data columns (total 81 columns):
 #   Column                               Non-Null Count   Dtype  
---  ------                               --------------   -----  
 0   ListingKey                           113937 non-null  object 
 1   ListingNumber                        113937 non-null  int64  
 2   ListingCreationDate                  113937 non-null  object 
 3   CreditGrade                          28953 non-null   object 
 4   Term                                 113937 non-null  int64  
 5   LoanStatus                           113937 non-null  object 
 6   ClosedDate                           55089 non-null   object 
 7   BorrowerAPR                          113912 non-null  float64
 8   BorrowerRate                         113937 non-null  float64
 9   LenderYield                          113937 non-null  float64
 10  EstimatedEffectiveYield              84853 non-null   float64
 11  EstimatedLoss

### Initial Notes on Dataset Structure:
>- The dataframe comprises **113,937** rows and **81** columns (features). 56 of these 81 columns (69%) contain numeric data. There is a wealth of information that describes the situations surrounding each loan in the dataset.

### Features of Interest
The dataset currently contains loads of information. However, the goal of this exploration is to understand the different borrower motivations when applying for loans, including the different factors that may influence loan favorability. As a result, we will direct our exploratory efforts towards the following features:
>(1.) **ListingCreationDate:** The date the listing was created.

>(2.) **ListingCategory (numeric):** The category of the listing that the borrower selected when posting their listing: 0 - Not Available, 1 - Debt Consolidation, 2 - Home Improvement, 3 - Business, 4 - Personal Loan, 5 - Student Use, 6 - Auto, 7- Other, 8 - Baby&Adoption, 9 - Boat, 10 - Cosmetic Procedure, 11 - Engagement Ring, 12 - Green Loans, 13 - Household Expenses, 14 - Large Purchases, 15 - Medical/Dental, 16 - Motorcycle, 17 - RV, 18 - Taxes, 19 - Vacation, 20 - Wedding Loans.

>(3.) **BorrowerRate:** The Borrower's interest rate for this loan.

>(4.) **BorrowerState:** The two letter abbreviation of the state of the address of the borrower at the time the Listing was created.

>(5.) **isBorrowerHomeowner:** A Borrower will be classified as a homowner if they have a mortgage on their credit profile or provide documentation confirming they are a homeowner.

>(6.) **IncomeRange:** The income range of the borrower at the time the listing was created.

>(7.) **IncomeVerifiable:** The borrower indicated they have the required documentation to support their income.

>(8.) **DebtToIncomeRatio:** The debt to income ratio of the borrower at the time the credit profile was pulled. This value is Null if the debt to income ratio is not available. This value is capped at 10.01 (any debt to income ratio larger than 1000% will be returned as 1001%).

>(9.) **StatedMonthlyIncome:** The monthly income the borrower stated at the time the listing was created.

>(10.) **ProsperRating (Alpha):** The Prosper Rating assigned at the time the listing was created between AA - HR. Applicable for loans originated after July 2009.

>(11.) **Term:** The length of the loan expressed in months.

>(12.) **EmploymentStatus:** The employment status of the borrower at the time they posted the listing.

>(13.) **EmploymentStatusDuration:** The length in months of the employment status at the time the listing was created.

>(14.) **Investors:** The number of investors that funded the loan.

>(15.) **LoanStatus:** The current status of the loan: Cancelled, Chargedoff, Completed, Current, Defaulted, FinalPaymentInProgress, PastDue. The PastDue status will be accompanied by a delinquency bucket.

>(16.) **LoanOriginalAmount:** The origination amount of the loan.

>(17.) **BorrowerAPR:** The Borrower's Annual Percentage Rate (APR) for the loan.

Generally, we aim to measure loan favorability in terms of **Prosper rating** and **Annual percentage rate (borrower APR)**.

## Data Assessment
---
We will now assess our dataframe, with particular focus on the features of interest. The major goal here is data exploration, hence findings on quality and tidiness will be collated under the assessment summary, and addressed using a brief and concise cleaning workflow.

First, let's create a list containing the key features for our analysis:

In [4]:
# Create a list of key feature names
key_features = ['ListingCreationDate', 'ListingCategory (numeric)', 'BorrowerRate', 'BorrowerState', 'IsBorrowerHomeowner',
                'IncomeRange', 'IncomeVerifiable', 'DebtToIncomeRatio', 'StatedMonthlyIncome', 'Term', 'ProsperRating (Alpha)',
                'EmploymentStatus', 'EmploymentStatusDuration', 'Investors', 'LoanStatus', 'LoanOriginalAmount', 'BorrowerAPR']

Next, we will examine a sample of records from these features in pandas, coupled with thorough visual assessments in a spreadsheet software like Ms Excel:

In [5]:
# Visually assess portions of the dataframe containing key features
df[key_features].sample(20)

ListingCreationDate  ListingCategory (numeric)  \
63069   2010-12-07 08:14:29.860000000                          7   
22065   2007-06-14 11:29:33.893000000                          0   
14683   2010-04-20 11:41:00.017000000                          3   
46657   2008-02-08 15:58:44.180000000                          4   
34597   2013-11-26 09:08:36.843000000                          1   
26849   2013-04-21 10:55:20.777000000                          1   
76110   2007-06-19 06:18:40.070000000                          0   
17482   2012-07-15 15:51:21.980000000                          1   
92422   2010-10-21 15:26:46.843000000                          1   
105296  2014-01-16 13:02:20.583000000                          1   
44741   2014-01-10 08:30:12.953000000                          1   
25266   2012-09-07 04:25:17.937000000                          7   
105382  2012-06-26 09:36:52.760000000                         20   
58030   2008-04-27 17:43:35.200000000                          1   
12454   2011-06-16 07:50:19.457000000                          1   
37414   2008-03-23 17:44:21.697000000                          4   
38992   2013-12-23 09:37:22.050000000                          7   
50936   2007-12-28 12:02:43.120000000                          1   
110938  2014-03-07 10:22:42.430000000                          1   
66752   2013-07-18 11:42:44.653000000                          2   

        BorrowerRate BorrowerState  IsBorrowerHomeowner     IncomeRange  \
63069         0.3500            MS                 True  $25,000-49,999   
22065         0.2100            MI                 True  $25,000-49,999   
14683         0.2625            OR                False       $1-24,999   
46657         0.1150            CO                 True  $50,000-74,999   
34597         0.2624            NC                False       $100,000+   
26849         0.2639            MA                 True  $25,000-49,999   
76110         0.2161            MO                False  $25,000-49,999   
17482         0.1902            FL                 True  $50,000-74,999   
92422         0.2750            NY                False       $100,000+   
105296        0.1355            NE                 True  $50,000-74,999   
44741         0.1535            OH                 True  $50,000-74,999   
25266         0.2524            SC                False  $25,000-49,999   
105382        0.2287            TX                False  $50,000-74,999   
58030         0.1140            WA                 True  $50,000-74,999   
12454         0.2699            WI                 True       $1-24,999   
37414         0.1960            MI                False  $25,000-49,999   
38992         0.1685            NY                 True  $25,000-49,999   
50936         0.1320            IL                False  $25,000-49,999   
110938        0.0874            MI                False       $100,000+   
66752         0.0869            MA                 True       $100,000+   

        IncomeVerifiable  DebtToIncomeRatio  StatedMonthlyIncome  Term  \
63069               True               0.58          2583.333333    36   
22065               True               0.05          3916.666667    36   
14683              False                NaN          1166.666667    36   
46657               True               0.19          5654.166667    36   
34597               True               0.03         12500.000000    36   
26849               True               0.30          3321.750000    36   
76110               True               0.28          2873.333333    36   
17482               True               0.28          5983.333333    36   
92422               True               0.06          9012.500000    36   
105296              True               0.37          4166.666667    36   
44741               True               0.30          4666.666667    36   
25266               True               0.03          3333.333333    36   
105382              True               0.19        

Finally, we will check for duplicates, then preview a quick summary of these features, to get an idea of the overall distribution of datatypes, null values and the nomenclature of the column names:

In [6]:
# Examine for duplicates
duplicates = df[key_features].duplicated().sum()
print('There are {} duplicate records in the dataset'.format(duplicates))

There are 871 duplicate records in the dataset


In [7]:
df[key_features].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 113937 entries, 0 to 113936
Data columns (total 17 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   ListingCreationDate        113937 non-null  object 
 1   ListingCategory (numeric)  113937 non-null  int64  
 2   BorrowerRate               113937 non-null  float64
 3   BorrowerState              108422 non-null  object 
 4   IsBorrowerHomeowner        113937 non-null  bool   
 5   IncomeRange                113937 non-null  object 
 6   IncomeVerifiable           113937 non-null  bool   
 7   DebtToIncomeRatio          105383 non-null  float64
 8   StatedMonthlyIncome        113937 non-null  float64
 9   Term                       113937 non-null  int64  
 10  ProsperRating (Alpha)      84853 non-null   object 
 11  EmploymentStatus           111682 non-null  object 
 12  EmploymentStatusDuration   106312 non-null  float64
 13  Investors                  11

### Assessment Summary
Visual and programmatic assessments, with regards to the features of interest, yeilded the following observations:
> 1. Useful features need to be isolated from the dataset.
> 2. **ListingCreationDate** is stored with the wrong datatype.
> 3. The numeric information in **ListingCategory (numeric)** could be better expanded to reflect the actual reasons for the loan. The data dictionary contains helpful information for this.
> 4. Longitude and Latitude information can be collected for **BorrowerState** to make visualizations easier.
> 5. _Not employed_ entries in **IncomeRange** could be safely replaced with `$0`.
> 6. Key features for loan favorability (**BorrowerAPR** and **ProsperRating (Alpha)**) contain null values.
> 7. **ListingCategory (numeric)** and **ProsperRating (Alpha)** can be reassigned with column names that are easier to work with.
> 8. There are 871 duplicate records in the dataset.

## Data Cleaning
---